sdk/storage/src/shares/share_directory_client.cpp

```diff
@@ -0,0 +1,272 @@
+    }
+    policies.emplace_back(std::make_unique<CommonHeadersRequestPolicy>());
+    policies.emplace_back(
+        std::make_unique<Core::Credentials::Policy::BearerTokenAuthenticationPolicy>(
+            credential, Azure::Storage::Details::c_StorageScope));
+    policies.emplace_back(std::make_unique<Azure::Core::Http::TransportPolicy>(
+        std::make_shared<Azure::Core::Http::CurlTransport>()));
+    m_pipeline = std::make_shared<Azure::Core::Http::HttpPipeline>(policies);
+  }
+
+  DirectoryClient::DirectoryClient(
+      const std::string& shareDirectoryUri,
+      const DirectoryClientOptions& options)
+      : m_shareDirectoryUri(shareDirectoryUri)
+  {
+    std::vector<std::unique_ptr<Azure::Core::Http::HttpPolicy>> policies;
+    policies.emplace_back(std::make_unique<Azure::Core::Http::TelemetryPolicy>(
+        Azure::Storage::Details::c_FileServicePackageName, FileServiceVersion));
+    policies.emplace_back(std::make_unique<Azure::Core::Http::RequestIdPolicy>());
+    for (const auto& p : options.PerOperationPolicies)
+    {
+      policies.emplace_back(p->Clone());
+    }
+    policies.emplace_back(
+        std::make_unique<Azure::Core::Http::RetryPolicy>(Azure::Core::Http::RetryOptions()));
+    for (const auto& p : options.PerRetryPolicies)
+    {
+      policies.emplace_back(p->Clone());
+    }
+    policies.emplace_back(std::make_unique<CommonHeadersRequestPolicy>());
+    policies.emplace_back(std::make_unique<Azure::Core::Http::TransportPolicy>(
+        std::make_shared<Azure::Core::Http::CurlTransport>()));
+    m_pipeline = std::make_shared<Azure::Core::Http::HttpPipeline>(policies);
+  }
+
+  DirectoryClient DirectoryClient::GetSubDirectoryClient(const std::string& subDirectoryName) const
+  {
+    auto builder = m_shareDirectoryUri;
+    builder.AppendPath(subDirectoryName, true);
+    return DirectoryClient(builder, m_pipeline);
+  }
+
+  FileClient DirectoryClient::GetFileClient(const std::string& filePath) const
+  {
+    auto builder = m_shareDirectoryUri;
+    builder.AppendPath(filePath, true);
+    return FileClient(builder, m_pipeline);
+  }
+
+  Azure::Core::Response<CreateDirectoryResult> DirectoryClient::Create(
+      const CreateDirectoryOptions& options) const
+  {
+    auto protocolLayerOptions = ShareRestClient::Directory::CreateOptions();
+    protocolLayerOptions.Metadata = options.Metadata;
+    protocolLayerOptions.FileAttributes = FileAttributesToString(options.SmbProperties.Attributes);
+    if (protocolLayerOptions.FileAttributes.empty())
+    {
+      protocolLayerOptions.FileAttributes = FileAttributesToString(FileAttributes::Directory);
+    }
+    protocolLayerOptions.FileCreationTime = options.SmbProperties.FileCreationTime;
+    protocolLayerOptions.FileLastWriteTime = options.SmbProperties.FileLastWriteTime;
+    if (options.DirectoryPermission.HasValue())
+    {
+      protocolLayerOptions.FilePermission = options.DirectoryPermission.GetValue();
```

Comment:
still, if in convenience layer one variable is nullable, why isn't it nullable in protocol layer?

Reply:
Not necessary, but this one should be changed.